Make GPS and Position Monitor CDU pages dynamic, cosmetics/realism

lhoenig opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->


**Summary of Changes**
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
This PR makes the GPS and Position Monitor pages of the CDU Data section dynamic, meaning they will continuously update to display the current values. It also includes realism and cosmetics improvements for these pages.


**Screenshots (if necessary)**
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->


**Additional context**
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):
lukas_#6140

DoToAdventures commented:
Hello, can you include screenshots as well as references used for these changes?

lhoenig commented:
Btw, the same trick I figured out to allow dynamic CDU pages would also be needed in A320_Neo_CDU_NavRadioPage.js, I tested it to no ill effect but I cannot see why that section should even be dynamic, so not including it in the PR for now.

lhoenig commented:
This is what the Position Monitor page looks like after the changes.
Before, it would say "MIXIRS" without the blank on the frozen page.
Also I added the "at HHMM" part to the title of the frozen page. Both changes according to http://www.a320dp.com/A320_DP/nav-flight-management/sys-14.7.2.html and http://www.a320dp.com/A320_DP/nav-flight-management/sys-14.7.2B.html
I hope that page is legit?
Oh, and I made the line with the three NAV values green on the unfrozen page, too. It was white before. Also per the same reference. The GPS section is visually unchanged, but also updates continuously. I couldnt't actually find a source saying explicitly they both do the updating, but it highly makes sense to me, also given the existence of the "Freeze" button.
![sn1](https://user-images.githubusercontent.com/4353030/93724164-f614fb80-fba4-11ea-84a2-981ce07330dc.PNG)
![sn2](https://user-images.githubusercontent.com/4353030/93724165-f6ad9200-fba4-11ea-83fe-38a43272cd77.PNG)

DoToAdventures commented:
Looks good, however I'm not sure about the source used. The A320DP site is based on CEOs, not NEOs and there might be MCDU changes there as well as the source is quite dated, we'll need an IRL pilot to confirm these changes, I think.

camilo625 commented:
Works as described

DoToAdventures reviewed (approved):
Ares confirmed this looks correct as it does in the real aircraft.

lhoenig commented:
Nice! In the commit I just added, I did the same thing for the IRS monitor pages, making them dynamic. I also added the frozen page, for which the button already existed. As a source I again only have the A320DP site (http://www.a320dp.com/A320_DP/nav-flight-management/sys-14.7.3A.html), but logically it should really be the same situation here.
Screenshot of new frozen page attached.
![irs](https://user-images.githubusercontent.com/4353030/93774377-809b4080-fc21-11ea-9e46-006966f5bd9e.PNG)

DoToAdventures commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSStatusFrozen.js:
```diff
@@ -0,0 +1,45 @@
+        let THDG = SimVar.GetSimVarValue("GPS GROUND TRUE HEADING", "radians") || "000";
+        let TTRK = SimVar.GetSimVarValue("GPS GROUND MAGNETIC TRACK", "radians") || "000";
+        let MHDG = SimVar.GetSimVarValue("GPS GROUND TRUE TRACK", "radians") || "000";
+        let WIND_DIR = SimVar.GetSimVarValue("AMBIENT WIND DIRECTION", "Degrees") || "000";
```
Getting ambient wind seems to cause the windindicator to jump around **a lot**. Perhaps implementation via aircraft wind is wiser? See the implementation of this on the ND in A32NX_NDInfo.js where simplane values are used instead of simvars.

lukecologne replied:
``` javascript
function getWindDirection() {
        var angle = SimVar.GetSimVarValue("AMBIENT WIND DIRECTION", "Degrees");
        return angle;
    }
    Simplane.getWindDirection = getWindDirection;
```
This is the Simplane function for getting the wind direction.
Simplane is most of the time (not always) just a wrapper for SimVar to make your life easier, it should be exactly the same in this case.

DoToAdventures replied:
Mh, I agree with you but I don't really understand where the discrepancy between the two values comes from then. The severity of this however seems to depend on the exact weather conditions, it's more noticable with stronger winds. 

https://streamable.com/0iozx7

DoToAdventures reviewed (changes requested):
After some talking with Ares, on the Position Monitor page FMGC needs to be changed too FMS1 and FMS2, the rest varies on the specific systems used.
![unknown](https://user-images.githubusercontent.com/40623152/93811050-d8e93700-fc4f-11ea-802f-87d2af7a75c0.png)

lhoenig commented:
Changed  FMGC to FMS as requested.

Also changed the Lat Long strings to fit the picture, see commit.

I also found out the reason the wind direction is displayed differently on the ND, they use some elaborate logic to smooth the angle, see
 `A32NX_NDInfo.js:setWind(_windAngle, _windStrength, _planeAngle, _force = false)`
which gets called in the update function using 
`this.setWind(Math.round(Simplane.getWindDirection()), Math.round(Simplane.getWindStrength()), Simplane.getHeadingMagnetic());`
Should I integrate that logic into the monitor pages?

DoToAdventures commented:
Tested, looking good! 
I'd suggest integrating the smoothing if possible, yes.

Externoak commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_GPSMonitor.js:
```diff
@@ -1,6 +1,11 @@
+    static ShowPage(mcdu, merit1, merit2, sat1, sat2) {
+        let currPos = new LatLong(SimVar.GetSimVarValue("GPS POSITION LAT", "degree latitude"), 
+                                  SimVar.GetSimVarValue("GPS POSITION LON", "degree longitude")).toShortDegreeString();
+        let currPosSplit = currPos.split("N")
```
It is me or N is not a constant value?
It could be N or S depending on the position of the plane in the world.

Externoak replied:
It is me or N is not a constant value?
It could be N or S depending on the position of the plane in the world.

Externoak commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSStatus.js:
```diff
@@ -1,7 +1,16 @@
+        SimVar.SetSimVarValue("L:FMC_UPDATE_CURRENT_PAGE", "number", 1);
+        let currPos = new LatLong(SimVar.GetSimVarValue("GPS POSITION LAT", "degree latitude"),
+                                  SimVar.GetSimVarValue("GPS POSITION LON", "degree longitude")).toShortDegreeString();
+        let currPosSplit = currPos.split("N")
```
Same as above.

lhoenig replied:
Oops, correct! I'll fix that later today!

lhoenig replied:
Resolved

Externoak commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PositionMonitorPage.js:
```diff
@@ -1,19 +1,28 @@
+        SimVar.SetSimVarValue("L:FMC_UPDATE_CURRENT_PAGE", "number", 1);
+        let currPos = new LatLong(SimVar.GetSimVarValue("GPS POSITION LAT", "degree latitude"),
+                                  SimVar.GetSimVarValue("GPS POSITION LON", "degree longitude")).toShortDegreeString();
+        let currPosSplit = currPos.split("N")
```
Same as above

Externoak reviewed (changes requested):
It is me or N is not a constant value?
It could be N or S depending on the position of the plane in the world.

lhoenig commented:
Correct handling of souther hemisphere position added and wind direction smoothing like on the ND implemented.
The smoothing depends on the delta time, which depends on the framerate, which can be slightly different for instruments.
That means the wind direction values on the MCDU will sometimes have a very slight deviation from the ND values. But it is like 1 degree for half a second.
It could only be made perfect if I could read the value directly from the ND, but I'm not sure how to do this. Does this suffice? I suggest you test it to see the effect for yourself. Its much better than before nonetheless.

DoToAdventures commented:
Much better!
